### ISSUE TITLE
Issues with resources when compiling

### ISSUE THREAD

**Issue by neuromantik33** [gradle-clojure/gradle-clojure#36]
Hello,

Firstly great work so far on the plugin, I'm happy testing it out on my projects and give some feedback.

That said, there is one bug that I think I can workaround with gradle must merits better handling. So I'm performing an AOT compilation and when evaluating a namespace I read a resource file, ex:

```clojure
(def ^:private schema
  (-> "graphql-schema.edn"
      io/resource
      slurp
      edn/read-string
      (lau/attach-resolvers resolvers)
      (lau/attach-scalar-transformers scalar-xformers)
      las/compile))
```

In my build.gradle I'm basing it off a lein build:

```gradle
sourceSets {
    dev {
        clojure.srcDirs = ['dev']
        resources.srcDirs = ['dev-resources']
    }
    main {
        clojure.srcDirs = ['src']
        resources.srcDirs = ['resources']
    }
    test {
        clojure.srcDirs = ['test']
    }
}
```

Well it seems that since gradle separates classes and resources it can't seem to locate the file, maybe a classpath issue. I can create a test project with exactly this behaviour, when building I get the following error:

```bash
nestrada@OLP00002145:~/repos/services/punchcard$ gradle clean build
Starting a Gradle Daemon, 1 stopped Daemon could not be reused, use --status for details

> Task :compileClojure
Creating fresh Clojure runtime.
SLF4J: Failed to load class "org.slf4j.impl.StaticMDCBinder".
SLF4J: Defaulting to no-operation MDCAdapter implementation.
SLF4J: See http://www.slf4j.org/codes.html#no_static_mdc_binder for further details.
ERROR:  clojure.lang.Compiler.CompilerException   java.lang.IllegalArgumentException: Cannot open <nil> as a Reader., compiling:(graphql.clj:74:7)


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileClojure'.
> A failure occurred while executing gradle_clojure.plugin.internal.ClojureWorker
   > java.lang.IllegalArgumentException: Cannot open <nil> as a Reader., compiling:(graphql.clj:74:7)

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

* Get more help at https://help.gradle.org

BUILD FAILED in 33s
3 actionable tasks: 3 executed
```

If you have any suggestions and what I can do to mitigate this issue (and reading from resources statically in a namespace is done quite often so I would prefer not changing it) I would much appreciate it.

Otherwise I'll be happy to help in any way I can, I really prefer this to boot as I do a lot of mixed clojure/groovy these days.

**Comment by ajoberstar** [comment]
I believe you could get this to work by doing the following:

```groovy
clojureCompile {
  source sourceSets.main.resources
}
```
[Looks like](https://github.com/gradle-clojure/gradle-clojure/blob/master/modules/gradle-clojure-plugin/src/main/java/gradle_clojure/plugin/ClojureBasePlugin.java#L61) we have the `ClojureCompile` tasks configured to just have the Clojure source on the classpath (and your compile dependencies). This appears to be the way [Gradle does this for Java](https://github.com/gradle/gradle/blob/master/subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java#L168), as well.

As for whether the default behavior should change, I'd want some of the other @gradle-clojure/developers to chime in. Not familiar enough with this use case to know how common it is. My first inclination would be to leave it as-is to align with other Gradle plugins.

**Comment by neuromantik33** [comment]
Unfortunately this seems to create an odd behavior with shimdandy where all resources are evaluated as clojure files, ex it seems to think that my logback.xml file is a namespace of sort. For example, after setting this option, I'm getting the following errors
```
LF4J: Failed to load class "org.slf4j.impl.StaticMDCBinder".
SLF4J: Defaulting to no-operation MDCAdapter implementation.
SLF4J: See http://www.slf4j.org/codes.html#no_static_mdc_binder for further details.
ERROR:  java.io.FileNotFoundException   Could not locate logback__init.class or logback.clj on classpath.

:compileClojure (Thread[Task worker for ':',5,main]) completed. Took 28.641 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileClojure'.
> A failure occurred while executing gradle_clojure.plugin.internal.ClojureWorker
   > Could not locate logback__init.class or logback.clj on classpath.

* Try:
Run with --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':compileClojure'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:100)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:63)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:88)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:197)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:107)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:124)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:80)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:105)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:99)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:625)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:580)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:99)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: org.gradle.workers.internal.DefaultWorkerExecutor$WorkExecutionException: A failure occurred while executing gradle_clojure.plugin.internal.ClojureWorker
        at org.gradle.workers.internal.DefaultWorkerExecutor$2.waitForCompletion(DefaultWorkerExecutor.java:156)
        at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForItemsAndGatherFailures(DefaultAsyncWorkTracker.java:99)
        at org.gradle.internal.work.DefaultAsyncWorkTracker.access$000(DefaultAsyncWorkTracker.java:34)
        at org.gradle.internal.work.DefaultAsyncWorkTracker$2.run(DefaultAsyncWorkTracker.java:83)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withoutLocks(DefaultWorkerLeaseService.java:196)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withoutProjectLock(DefaultWorkerLeaseService.java:145)
        at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForCompletion(DefaultAsyncWorkTracker.java:79)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:130)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:197)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:107)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:111)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:92)
        ... 27 more
Caused by: java.io.FileNotFoundException: Could not locate logback__init.class or logback.clj on classpath.
        at clojure.lang.RT.load(RT.java:463)
        at clojure.lang.RT.load(RT.java:426)
        at clojure.core$load$fn__6548.invoke(core.clj:6046)
        at clojure.core$load.invokeStatic(core.clj:6045)
        at clojure.core$load.doInvoke(core.clj:6029)
        at clojure.lang.RestFn.invoke(RestFn.java:408)
        at clojure.core$load_one.invokeStatic(core.clj:5848)
        at clojure.core$compile$fn__6553.invoke(core.clj:6056)
        at clojure.core$compile.invokeStatic(core.clj:6056)
        at clojure.core$compile.invoke(core.clj:6048)
        at gradle_clojure.tools.clojure_compiler$compiler$fn__163.invoke(clojure_compiler.clj:32)
        at gradle_clojure.tools.clojure_compiler$compiler.invokeStatic(clojure_compiler.clj:25)
        at gradle_clojure.tools.clojure_compiler$compiler.invoke(clojure_compiler.clj:23)
        at clojure.lang.AFn.applyToHelper(AFn.java:165)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.lang.Var.applyTo(Var.java:702)
        at clojure.core$apply.invokeStatic(core.clj:657)
        at clojure.core$apply.invoke(core.clj:652)
        at clojure.lang.Var.invoke(Var.java:385)
        at org.projectodd.shimdandy.impl.ClojureRuntimeShimImpl.invoke(ClojureRuntimeShimImpl.java:124)
        at org.projectodd.shimdandy.impl.ClojureRuntimeShimImpl.invoke(ClojureRuntimeShimImpl.java:117)
        at gradle_clojure.plugin.internal.ClojureWorker.run(ClojureWorker.java:60)
        at org.gradle.workers.internal.WorkerServer.execute(WorkerServer.java:31)
        at org.gradle.workers.internal.WorkerDaemonServer.execute(WorkerDaemonServer.java:38)
        at org.gradle.workers.internal.WorkerDaemonServer.execute(WorkerDaemonServer.java:25)
        at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:98)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:146)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:128)
        at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
        ... 3 more
```
I removed the logback.xml just for testing, and my resources directory contains a datomic-schema.edn file, and this is what I get now:
```
> Task :compileClojure
Creating fresh Clojure runtime.
ERROR:  java.io.FileNotFoundException   Could not locate datomic_schema__init.class or datomic_schema.clj on classpath. Please check that namespaces with dashes use underscores in the Clojure file name.


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileClojure'.
> A failure occurred while executing gradle_clojure.plugin.internal.ClojureWorker
   > Could not locate datomic_schema__init.class or datomic_schema.clj on classpath. Please check that namespaces with dashes use underscores in the Clojure file name.
```
Maybe by fiddling with the main source set classpath rather than setting the resources as source is what is needed however I'm open to suggestions.

**Comment by ajoberstar** [comment]
Ah, right [our compile task](https://github.com/gradle-clojure/gradle-clojure/blob/master/modules/gradle-clojure-plugin/src/main/java/gradle_clojure/plugin/tasks/ClojureCompile.java#L125) isn't doing additional checks for which files are in the source, it just assumes they're all Clojure and infers namespaces for every file.

Maybe this would work:

```groovy
clojureCompile {
  classpath += sourceSets.main.resources
}
```

**Comment by jszakmeister** [comment]
I can't remember for certain off the top of my head, but I did think that the resource paths were expected to be added to the classpath.  I don't know if Gradle does this by default (we've not been fortunate enough to use Gradle is our work place just yet--we're eyeing it for polyglot stuff), but I'm pretty sure Leiningen does.  The Gradle docs don't really say what is done in the Java tasks. :-(

**Comment by ajoberstar** [comment]
Gradle (for Java projects) does not put resources on the classpath of Java compilation. It probably isn't useful for Java, like it could be for Clojure.

With Gradle, resources always have a "process" task where (as opposed to compile for source code) you can do some filtering or substitution on the files. If we did put them on the classpath (by default) for Clojure we may want to use the ones coming out of that process task instead of the original source dir.

**Comment by ajoberstar** [comment]
Similar to #40, we might be able to follow the approach Groovy uses to put Java on the classpath to do the same with resources.

**Comment by jszakmeister** [comment]
> Gradle (for Java projects) does not put resources on the classpath of Java compilation. It probably isn't useful for Java, like it could be for Clojure.

Yeah, I was going to say I don't think they do for compilation, but I believe they do for runtime-related stuff.  But your right, since compilation and execution are pretty much the same thing in Clojure, we might need to for Clojure.

**Comment by ajoberstar** [comment]
The fix is part of 0.3.0. Please reopen if this didn't fix the issue.
